toolchest_client/tools/tool.py

```diff
@@ -322,25 +339,47 @@ def _check_thread_health(self):
                 self._kill_query_threads()
                 raise ToolchestException("A job irrecoverably failed. See logs above for details.")
 
-    def _wait_for_threads_to_finish(self, check_health=True, wait_for_upload=True):
-        """Waits for all jobs and their corresponding threads to finish while printing their statuses."""
-        elapsed_seconds = 0
+    async def _wait_for_threads_to_finish(self):
+        """Waits for all jobs and their corresponding threads to finish while printing their statuses
+        or streaming thread output, if enabled."""
         # Wait for all threads to finish uploading
-        if wait_for_upload:
-            self._wait_for_threads_to_upload()
+        self._wait_for_threads_to_upload()
 
         for thread in self.query_threads:
             increment_seconds = 5
             while thread.is_alive():
-                if check_health:
-                    self._check_thread_health()
-                self._pretty_print_pipeline_segment_status(elapsed_seconds)
-                elapsed_seconds += increment_seconds
-                time.sleep(increment_seconds)
+                self._check_thread_health()
+
+                if self.streaming_enabled and self.streaming_client.params_initialized:
+                    # Initialize and start streaming (in place of regular status updates).
+                    if not self.streaming_task:
+                        # Clear the last status line before streaming begins
+                        print("Printed lines:".ljust(120))
+                        self.streaming_task = asyncio.create_task(self.streaming_client.receive_stream())
+                    # Check if an exception was raised in the streaming task
+                    elif self.streaming_task.done():
+                        if self.streaming_task.exception():
+                            self._kill_query_threads()
+                            error_message = (
+                                "A streaming error was encountered. See logs above for details.\n"
+                                "Try re-running the same Toolchest command with this argument:\n"
+                                "\tstreaming_enabled=False"
+                            )
+                            raise ToolchestException(error_message) from self.streaming_task.exception()
+                else:
+                    self._pretty_print_pipeline_segment_status()
+
+                self.elapsed_seconds += increment_seconds
+                await asyncio.sleep(increment_seconds)
+
+            # Wait for the stream to complete before printing final status line
+            if self.streaming_task:
+                await self.streaming_task
+
```

Comment:
This is very nested, is there a cleaner way to do this?

Reply:
Not by much, but I've removed some of the nesting.

Reply:
More clear now! It looks like this is no PR comment is longer attached to the block of code, but responding here for continuity,

There's always a way to make things less nested: moving the bulk of the nested logic into its own function (e.g. how `_check_thread_health` works), which could make this more readable.

As-is, the function is good, but thought I'd mention that there is still a clear path toward removing nesting.